### PR TITLE
Bake RESET animation.

### PR DESCRIPTION
**Pull request by fire**
Co-Authored-by MMMaellon
Co-Authored-by Eron.

Corresponding pr to https://github.com/godotengine/godot-proposals/issues/2961.

### CODE REVIEW

**Comment by akien-mga** [comment]
> Co-Authored-by MMMaellon
Co-Authored-by Eron.

Note that if you want them properly credited as authors by GitHub, you have to use this exact syntax (and IIRC at the end of the commit message):

```
Co-authored-by: Name <email>
```

File: editor/import/resource_importer_scene.h [hunk 1]
```diff
@@ -32,6 +32,7 @@
 #define RESOURCEIMPORTERSCENE_H
 
 #include "core/io/resource_importer.h"
+#include "editor/import/editor_importer_bake_reset.h"
```
Comment (akien-mga) [inline editor/import/resource_importer_scene.h:35]: This is used in the cpp, so this should go in the cpp.

File: editor/import/editor_importer_bake_reset.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "editor/import/editor_importer_bake_reset.h"
+#include "core/error/error_macros.h"
```
Comment (akien-mga) [inline editor/import/editor_importer_bake_reset.cpp:32]:
```suggestion
#include "editor/import/editor_importer_bake_reset.h"

#include "core/error/error_macros.h"
```

File: editor/import/editor_importer_bake_reset.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+#ifndef RESOURCE_IMPORTER_BAKE_RESET_H
+#define RESOURCE_IMPORTER_BAKE_RESET_H
+#include "core/math/transform_3d.h"
+#include "editor/import/scene_importer_mesh_node_3d.h"
+#include "resource_importer_scene.h"
+#include "scene/3d/mesh_instance_3d.h"
+#include "scene/3d/node_3d.h"
+#include "scene/3d/skeleton_3d.h"
+#include "scene/animation/animation_player.h"
```
Comment (akien-mga) [inline editor/import/editor_importer_bake_reset.h:39]: I can see several headers here which are not used in this file.

You should always put the includes where they're used. If only the `.cpp` file uses them, then they don't need to be in the header. The more includes you have in headers, the longer the engine build times, as this header (and all files that include it) would always be rebuild whenever one of those dependency changes.

While if only the `.cpp` needs those, only the `.cpp` gets rebuilt when they change.

File: editor/import/editor_importer_bake_reset.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef RESOURCE_IMPORTER_BAKE_RESET_H
+#define RESOURCE_IMPORTER_BAKE_RESET_H
+#include "core/math/transform_3d.h"
```
Comment (akien-mga) [inline editor/import/editor_importer_bake_reset.h:33]:
```suggestion
#ifndef RESOURCE_IMPORTER_BAKE_RESET_H
#define RESOURCE_IMPORTER_BAKE_RESET_H

#include "core/math/transform_3d.h"
```

File: editor/import/resource_importer_scene.cpp [hunk 5]
```diff
@@ -1411,6 +1412,11 @@ Error ResourceImporterScene::import(const String &p_source_file, const String &p
 	_pre_fix_node(scene, scene, collision_map);
 	_post_fix_node(scene, scene, collision_map, scanned_meshes, node_data, material_data, animation_data, fps);
 
+	bool use_bake_reset_animation = p_options["animation/bake_reset_animation"];
+	if (use_bake_reset_animation) {
+		BakeReset bake_reset;
+		bake_reset._bake_animation_pose(scene, "RESET");
+	}
 	String root_type = p_options["nodes/root_type"];
```
Comment (akien-mga) [inline editor/import/resource_importer_scene.cpp:1420]:
```suggestion
	bool use_bake_reset_animation = p_options["animation/bake_reset_animation"];
	if (use_bake_reset_animation) {
		BakeReset bake_reset;
		bake_reset._bake_animation_pose(scene, "RESET");
	}

	String root_type = p_options["nodes/root_type"];
```

File: editor/import/editor_importer_bake_reset.cpp [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+			r_rest_bones[StringName(String(skeleton->get_owner()->get_path_to(skeleton)) + ":" + bone_name)] = rest_bone;
+			break;
+		}
+	}
+}
+
+void BakeReset::_fix_skeleton(Skeleton3D *p_skeleton, Map<StringName, BakeReset::BakeResetRestBone> &r_rest_bones) {
+	int bone_count = p_skeleton->get_bone_count();
+
+	//First iterate through all the bones and create a RestBone
```
Comment (akien-mga) [inline editor/import/editor_importer_bake_reset.cpp:168]: Please use this style for comments (space after `//`, capital letter, final dot):
```suggestion
	// First iterate through all the bones and create a RestBone.
```

File: editor/import/editor_importer_bake_reset.h (hunk 3 above)
Comment (fire) [reply editor/import/editor_importer_bake_reset.h]: Forward declared and only using core_binds now.

**Comment by fire** [comment]
Sample project for verification. https://drive.google.com/drive/folders/15TnICjTpcr3dGEAJwcuWb69B5Gy9HP4f from @AndreaCatania

**Comment by RandomShaper** [comment]
Code-wise I don't see any big issue and it looks as it does what it promises.

If anything, I'd add a comment with a bird view explanation of the algorithm, like a numered bullet list) in `editor_importer_bake_reset.cpp`, even including some URLs of the reference materials in the proposal. The goal is to make this somehow obscure piece of code not so intimidating for potential future maintainers.

**Comment by fire** [comment]
I have copy and pasted the proposal details into the source file.

**Comment by RandomShaper** [comment]
Cool, but now I'd say it's too unscoped. I'm sorry to annoy you with this, but it'd be better to include an explanation about what each step of the algorithm do (which may include references to the implementation it has been inspired in), instead of the rationale for adding the feature.

File: editor/import/editor_importer_bake_reset.cpp [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+#include "editor/import/editor_importer_bake_reset.h"
+#include "core/math/transform_3d.h"
+#include "editor/import/scene_importer_mesh_node_3d.h"
+#include "resource_importer_scene.h"
+#include "scene/3d/mesh_instance_3d.h"
+#include "scene/3d/node_3d.h"
+#include "scene/3d/skeleton_3d.h"
+#include "scene/animation/animation_player.h"
+
+#include "core/error/error_macros.h"
```
Comment (akien-mga) [inline editor/import/editor_importer_bake_reset.cpp:40]:
```suggestion
#include "editor/import/editor_importer_bake_reset.h"

#include "core/error/error_macros.h"
#include "core/math/transform_3d.h"
#include "editor/import/scene_importer_mesh_node_3d.h"
#include "resource_importer_scene.h"
#include "scene/3d/mesh_instance_3d.h"
#include "scene/3d/node_3d.h"
#include "scene/3d/skeleton_3d.h"
#include "scene/animation/animation_player.h"
```

File: editor/import/editor_importer_bake_reset.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef RESOURCE_IMPORTER_BAKE_RESET_H
+#define RESOURCE_IMPORTER_BAKE_RESET_H
+
+#include "core/core_bind.h"
```
Comment (akien-mga) [inline editor/import/editor_importer_bake_reset.h:34]: Why `core_bind.h`?

Comment (fire) [reply editor/import/editor_importer_bake_reset.h:34]: I could directly import String and StringName, and node but it was included in core bind.

Comment (akien-mga) [reply editor/import/editor_importer_bake_reset.h:34]: You should import what you use. You're definitely not using core binds.
And almost none of the engine does either:
```
$ rg '#include "core/core_bind.h"'
modules/fbx/data/fbx_material.cpp
35:#include "core/core_bind.h"

core/register_core_types.cpp
35:#include "core/core_bind.h"

core/config/project_settings.cpp
33:#include "core/core_bind.h" // For Compression enum.
```

The `fbx` one is actually a bad practice, I put a FIXME.

Comment (fire) [reply editor/import/editor_importer_bake_reset.h]: I have imported node.h for the same purpose. The Node class is directly used.

**Comment by fire** [comment]
@RandomShaper I added some comments at crucial parts of the algorithm, but it doesn't seem much.

**Comment by RandomShaper** [comment]
I think it does the trick quite well. :)

**Review by RandomShaper — APPROVED** [review]
Looks good to me. My only doubt is about how much green light we can consider the proposal got. For my part, I understand this feature aligns very well with the Godot principle of parallel work among a team and the implementation is clean.

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
If the feature is wanted in `3.x` too, this might need a dedicated backport, as the `resource_importer_scene.cpp` code changed significantly.

**Comment by fire** [comment]
I'll see what I can do.

**Comment by slapin** [comment]
Could this be backported to 3.x please?

**Pull request merged**